### ISSUE TITLE
What is the definition of a "valid event"

### ISSUE THREAD

**Issue by raunraun** [KhronosGroup/OpenCL-Docs#482]
The phrase "...valid events" is used in every API that accepts an event waitlist, but there is no definition of what a valid event is.

I think a valid event is any event that can be queried by clGetEventInfo - returned by an OpenCL API, is associated with an active context (if applicable), and has not been destroyed (if applicable).

**Comment by bashbaug** [comment]
This makes sense to me.  We should define "validity" for other objects too and not just events.

**Comment by kpet** [comment]
That's timely! I raised the question of object validity a few weeks ago in teleconferences.

Specifically: when are API calls required to cleanly reject invalid objects with `CL_INVALID_*` as opposed to being allowed undefined behaviour? There are at least three cases we should consider:
- `nullptr`, we should always require that implementations handle this case IMO
- pointers to a memory zone that is both big enough to contain an object of the type being considered and accessible
- pointers to memory that is not accessible (not mapped, etc)

There are negative CTS tests being written for these error cases so it would be good to agree on the scenarios that all  implementations must support.

**Comment by b-sumner** [comment]
I think nullptr should be it.  We shouldn't require implementations to perform more expensive checks.  I appreciate the fact that vulkan can put extensive (or even all) tests in layers.

**Comment by bashbaug** [comment]
We should keep the question of "what is a valid event" separate from "what do implementations need to check for and return deterministic errors".  In other words, I think it's useful to define "validity" even if we cannot reliably return an error in all cases.

That being said, are either of these cases common and are these checks feasible?

> * pointers to a memory zone that is both big enough to contain an object of the type being considered and accessible
> * pointers to memory that is not accessible (not mapped, etc)

IMHO the more common scenarios that produce an invalid object are:

1. Using an object that was either never allocated or that has been released and destroyed ("dangling pointer").
2. Using an object that is valid in one OpenCL context in another OpenCL context, either for the same OpenCL device and platform or for a different OpenCL device or platform.

Although I think it is possible to test for both of these scenarios the tests are expensive.  I agree it would be preferable to put these expensive checks in a validation layer, if possible.

**Comment by jlewis-austin** [comment]
> Using an object that is valid in one OpenCL context in another OpenCL context, either for the same OpenCL device and platform or for a different OpenCL device or platform.

For this one, the spec calls out specific checks, such as
`The context associated with events in event_wait_list and command_queue must be the same.`
This seems to imply that context checks are independent of validity checks; i.e. it's treating the object as valid without regard to context?
Of course this is from the "what do implementations need to check for" perspective with the current language. It does seem reasonable to define an object as invalid if it wanders outside its context.

**Comment by bashbaug** [comment]
> `The context associated with events in event_wait_list and command_queue must be the same.`
> This seems to imply that context checks are independent of validity checks; i.e. it's treating the object as valid without regard to context?

Yeah, I'm just observing that this is a case that currently could return an "invalid object" error code, specificaly `CL_INVALID_CONTEXT`, even though both the event object and the command queue are valid.  This is admittely a bit of a corner case and not relevant to the initial question of "how do you determine if an event in an event wait list is valid"?

Maybe it's helpful to enumerate the cases that _could_ be considered invalid and whether implementations _must_ return `CL_INVALID_EVENT_WAIT_LIST` in these scenarios.

So far I think we're in agreement that this case is an invalid event and must produce `CL_INVALID_EVENT_WAIT_LIST`:

1. Passing a `NULL` event.

Should these cases produce `CL_INVALID_EVENT_WAIT_LIST`?

2. Passing some other valid object that is not an event?  (Probably.)
3. Passing some other random bogus (but accessible) pointer?  (Probably?  How big does the accessible memory need to be?)
4. Passing an event from another context for the same platform/device?  (Not `CL_INVALID_EVENT_WAIT_LIST` but probably `CL_INVALID_CONTEXT`?)
5. Passing a previously free'd event?  (Does this depend if the memory is accessible?)
6. Passing a bogus and inaccessible pointer?

(2) and (3) seem doable and fairly cheap.  (4) is doable and only a bit more expensive.  (5) and (6) could be tough in at least some cases... (6) especially.

What other cases should we consider?

**Comment by jlewis-austin** [comment]
Thanks, that sounds like a good list. I think I might be missing your point on Item 4; that's the one that I'm reading as already required to return CL_INVALID_CONTEXT. Is there further nuance there or are you just listing it to consider whether the error is correct?
I haven't checked in a while but in the past we found some of these not possible to hit in a ICD implementation because the loader either caught the error first (due to item 1 null pointers mainly IIRC), or crashed before the driver had a chance to handle it (item 6). We'd have to double-check that to avoid requiring anything that's out of the driver's control.

**Comment by bashbaug** [comment]
Yeah, (4) was just listed for completeness, and I agree the spec is pretty clear that it should be `CL_INVALID_CONTEXT` in this case.  If we expand this discussion to other "valid" objects it'd be something to check though.

Good point about (3) also - I think this is the one that will likely crash if it is the "dispatchable object" for an ICD loader implementation, since the ICD dispatch table will also be bogus.

**Comment by alycm** [comment]
https://github.com/KhronosGroup/OpenCL-Docs/pull/834 is merged and covers the "**Definition of a valid object**" part of the above comment, but not the "**Requirements on implementations**" part.

**Comment by bashbaug** [comment]
Copying a PR comment here to document one additional improvement we should consider before closing this issue:

https://github.com/KhronosGroup/OpenCL-Docs/pull/842#pullrequestreview-1108450453

Short summary: There are several APIs where the passed-in platform can be `NULL` without this being considered an invalid platform and we need to ensure this is clear.
